### ISSUE TITLE
Better guidance for dealing with JPA transactions

### ISSUE THREAD

**Issue by gmethvin** [playframework/playframework#6489]
Reading through the [JPA documentation](https://www.playframework.com/documentation/2.5.x/JavaJPA), I don't think we're providing the right guidance in terms of how to execute blocking database calls. It should be very clear that you usually do not want to execute blocking calls in the default execution context.

Also, It seems to me that the `@Transactional` annotation is broken, since you end up running blocking database calls on Play's default execution context, and we strongly suggest that as the way to do it. We could fix this, but the execution context needs to come from somewhere: either Play needs to be responsible for creating it or the user needs to bind it.

If we decide to require a separate execution context, we could also provide an async JPA API, where methods like `withTransaction` should return `CompletionStage<T>` and automatically run the transaction in a separate execution context.

**Comment by marcospereira** [comment]
If I remember correctly, `@Transactional` was "inherited" from Play1.

> We could fix this, but the execution context needs to come from somewhere: either Play needs to be responsible for creating it or the user needs to bind it.

Maybe a default execution context declaration can be added to [play-java-jpa reference.conf declaration](https://github.com/playframework/playframework/blob/master/framework/src/play-java-jpa/src/main/resources/reference.conf).

> If we decide to require a separate execution context, we could also provide an async JPA API, where methods like withTransaction should return CompletionStage<T> and automatically run the transaction in a separate execution context.

We have to be cautious about thread locals here:

https://github.com/playframework/playframework/blob/master/framework/src/play-java-jpa/src/main/java/play/db/jpa/JPAEntityManagerContext.java

**Comment by gmethvin** [comment]
@marcospereira If we make an async API we could just disallow the use of the thread local context.  All the methods would accept `EntityManager => T` and pass the EntityManager directly.

The problem is just how to "fix" the `@Transactional` annotation, since there's no API to get the entity manager other than `JPA.em()`

**Comment by mkurz** [comment]
I got rid of the thread local in #6777.
Also I think I solved the problem for `JPA.em()` when using `@Transactional`  - now we always have to pass the context from which the entity manager should be retrieved via `JPA.em(ctx())`. In a request a context is always available. So no need for `Http.Context.current()` anymore.

We have had async `withTransaction` methods [in Play 2.4](https://github.com/playframework/playframework/blob/2.4.8/framework/src/play-java-jpa/src/main/java/play/db/jpa/DefaultJPAApi.java#L171) and below already, but it has been removed because it [was causing deadlocks](https://github.com/playframework/playframework/pull/4225#issuecomment-92597680). I think this was because it didn't use its own execution context. So by providing an default ec via the play-java-jpa `reference.conf` I think we could get it working. The user could then override this ec in application.conf or/and we could provide an overloaded `withTransactionAsync` where he/she can pass an ec explicitly.
WDYT?
